enhancements/baremetal/coreos-image-in-release.md

```diff
@@ -0,0 +1,282 @@
+* Update the CoreOS QCOW2 image in the cluster with each OpenShift release.
+* Provide CoreOS images for platforms other than baremetal.
+* Eliminate the extra reboot performed to update CoreOS after initial
+  provisioning.
+
+## Proposal
+
+Build a container image containing the latest CoreOS ISO and the
+`coreos-installer` binary. This container can be used e.g. as an init container
+to make the ISO available where required. The `coreos-installer iso extract
+pxe` command can also be used to produce the kernel, initrd, and rootfs from
+the ISO for the purposes of PXE booting.
+
+This image could either replace the existing content of the
+`ironic-ipa-downloader` repo, be built from the new
+`image-customization-controller` repo, or be built from a new repo.
+
+### User Stories
+
+As an operator of a disconnected cluster, I want to upgrade my cluster and have it to continue to work for provisioning baremetal machines.
+
+As an operator of an OpenShift cluster, I want to add to my cluster new
+hardware that was not fully supported in RHEL at the time I installed the
+cluster.
+
+As an operator of an OpenShift cluster, I want to ensure that the OS running on
+hosts prior to them being provisioned as part of the cluster is up to date with
+bug and security fixes.
+
+### Implementation Details/Notes/Constraints
+
+We will need to restore a [change to the Machine Config
+Operator](https://github.com/openshift/machine-config-operator/pull/1792) to
+allow working with different versions of Ignition that was [previously
+reverted](https://github.com/openshift/machine-config-operator/pull/2126) but
+should now be viable after [fixes to the
+installer](https://github.com/openshift/installer/pull/4413).
+
+The correct version of CoreOS to use is available from the [stream
+metadata](https://github.com/openshift/installer/blob/master/data/data/rhcos-stream.json)
+in the openshift-installer. This could be obtained by using the installer
+container image as a build image. Deriving the container image from the
+installer container has the convenient side-effect of always triggering a
+rebuild when the version changes.
+
+Because OpenShift container images are built in an offline environment, it is
+not possible to simply download the image from the public Internet at container
+build time. Instead this will be accomplished by *MAGIC*. Uploading the ISOs to
+the lookaside cache in Brew is one possibility.
+
+Different processor architectures require different ISOs (though all are
+represented in the stream metadata). It may make sense to have a separate
+Dockerfile for each architecture. For the foreseeable future, we only need to
+be able to pull a container for the local architecture that contains an ISO for
+that same architecture. In the future, it is possible that the containers for
+each architecture's ISO may themselves need to be multi-architecture, so that
+it is possible to run a container image containing an ISO for a *different*
+architecture.
```

Comment:
implementation details (I don't think we need to split image builds or anything fancy)

```suggestion
build time. Instead this will be accomplished by ART defining a "modifications"
hook that, each time the source is updated before a build, checks whether
`data/data/rhcos-stream.json` has changed. If it has, ISOs from the locations
given there are downloaded and made available in the build context with an
arch-specific name, something like `rhcos.$arch.iso` so that the Dockerfile can
select the appropriate arch.

Different processor architectures require different ISOs (though all are
represented in the stream metadata). For the foreseeable future, we only need to
be able to pull a container for the local architecture that contains an ISO for
that same architecture. In the future, it is possible that the containers for
each architecture's ISO may themselves need to be multi-architecture, so that
it is possible to run a container image containing an ISO for a *different*
architecture.
```